mvebu: add Kobol Helios 4 device

bobafetthotmail opened the pull request:
The Helios 4 is a NAS from Kobol
that is powered by an Armada 38x
MicroSOM from Solidrun, similarly
to Clearfog.

This device has:
-Armada 38x CPU
 (dual core ARMv7 1.6 Ghz)
-2 GB of ECC RAM
-Gigabit ethernet (Marvell)
-2x USB 3.0 ports
-4x Sata 3.0 ports
-i2c header (J9 |>GND|SDA|SCL|VCC)
-2x 3-pin fan headers with PWM
-micro-usb port is a TTL/UART to
 USB converter connected to TTL
-MicroSD card slot

To install, write the disk image
on a micro SD card with dd or
win32 disk imager, insert the
card in the slot.
Check that the dip switch battery
for boot selection is as follows
Switch 1 and 2 down/off, switches
3, 4, 5 up/on.

adrianschmutzler commented on target/linux/mvebu/cortexa9/base-files/etc/init.d/helios4_fancontrol:
```diff
@@ -0,0 +1,97 @@
+}
+
+boot(){
+start
```
Indent.

IIRC, start() will be triggered on boot automatically, so this additional block can be dropped.

bobafetthotmail replied:
ok

adrianschmutzler commented on target/linux/mvebu/image/cortexa9.mk:
```diff
@@ -232,3 +232,18 @@ define Device/solidrun_clearfog-pro-a1
+  DEVICE_PACKAGES := mkf2fs e2fsprogs partx-utils swconfig
+  IMAGES := sdcard.img.gz
+  IMAGE/sdcard.img.gz := boot-scr | boot-img-ext4 | sdcard-img-ext4 | gzip | append-metadata
+  DEVICE_DTS := armada-388-helios4
```
Should work with `SOC := armada-388` instead.

bobafetthotmail replied:
uhm, not sure what you mean. This device has a specific dts, and I think I need to specify its name here.

adrianschmutzler replied:
image/Makefile contains `DEVICE_DTS = $$(SOC)-$(lastword $(subst _, ,$(1)))`, so the DTS file name will be constructed from the device name and the SOC ...

bobafetthotmail replied:
ok

adrianschmutzler commented on target/linux/mvebu/image/cortexa9.mk:
```diff
@@ -232,3 +232,18 @@ define Device/solidrun_clearfog-pro-a1
+  IMAGES := sdcard.img.gz
+  IMAGE/sdcard.img.gz := boot-scr | boot-img-ext4 | sdcard-img-ext4 | gzip | append-metadata
+  DEVICE_DTS := armada-388-helios4
+  SUPPORTED_DEVICES += armada-388-helios4
```
What's that for?

bobafetthotmail replied:
I don't know, I've just copied the Clearfog since this device is very similar. Also all other devices in this file have a similar entry.

adrianschmutzler replied:
Then just drop this line. It's for adding _additional_ device names, either after a rename or for cross-flashing, which both doesn't apply here. The primary name is created from the device node name.

bobafetthotmail replied:
ok

adrianschmutzler commented on target/linux/mvebu/cortexa9/base-files/etc/init.d/helios4_fancontrol:
```diff
@@ -0,0 +1,97 @@
+
+start() {
+	case $(board_name) in
+	kobol,helios4)
```
I'm not sure whether the file should be named more generic, if devices are switched inside anyway.

bobafetthotmail replied:
yeah will give it a more generic name

bobafetthotmail committed:
uboot-mvebu: add uboot for helios 4

add u-boot for Helios 4 NAS

Signed-off-by: Alberto Bursi <bobafetthotmail@gmail.com>

bobafetthotmail commented:
updated with feedback. 
Moved helios4 entries in the makefiles to keep alphabetic order.
Added dts patch for status led alias (since this target dropped diag.sh script in favor of that) so the status led is used by OpenWrt.

adrianschmutzler commented on target/linux/mvebu/cortexa9/base-files/etc/init.d/fancontrol:
```diff
@@ -0,0 +1,93 @@
+		pwm_value=$pwm_step_10
+	elif [ $cpu_temp -gt $target_temp_10 ]; then
+		pwm_value=255
+	fi
```
Since elif will only enter one case, one should be able to reduce the conditions to one condition here for most cases:
```
	if [ $cpu_temp -le $cpu_temp_min ]; then
		pwm_value=0
	elif [ $cpu_temp -le $target_temp_1  ]; then
		pwm_value=$pwm_step_1
	elif [ $cpu_temp -le $target_temp_2 ]; then
		pwm_value=$pwm_step_2

...
```

adrianschmutzler replied:
And the final elif would just be an "else" then ...

adrianschmutzler commented on target/linux/mvebu/patches-4.19/560-helios4-dts-status-led-alias.patch:
```diff
@@ -0,0 +1,28 @@
+ 
+ 	system-leds {
+ 		compatible = "gpio-leds";
+-		status-led {
```
It's a nitpick, but I'd keep the designation "status" here:
`led_status: status-led {`

bobafetthotmail replied:
I copied how it was done for other mvebu devices https://github.com/openwrt/openwrt/blob/master/target/linux/mvebu/patches-5.4/240-linksys-status-led.patch

but ok, can change that

adrianschmutzler replied:
In the references, the LED has been named power beforehand, and that is kept. In your case, the LED is called "status" beforehand, and that should be kept.

adrianschmutzler commented on target/linux/mvebu/cortexa9/base-files/etc/init.d/fancontrol:
```diff
@@ -0,0 +1,93 @@
+	target_temp_7=$(( cpu_temp_min + ( temp_step * 7 ) ))
+	target_temp_8=$(( cpu_temp_min + ( temp_step * 8 ) ))
+	target_temp_9=$(( cpu_temp_min + ( temp_step * 9 ) ))
+	target_temp_10=$cpu_temp_max
```
I wonder whether all of that could be done much easier with a for loop (pseudo-code):

for i in 0:1:10
   if realtemp > $(( cpu_temp_min + ( temp_step * i ) ))
      continue
   fi
   pwm_value = $(( pwm_min + ( ( 255 - pwm_min ) / 10 * i ) ))
   break
fi

bobafetthotmail replied:
the idea was to precompute the values so it would not need to do the same calculation over and over again every time it needs to adjust the temp.

adrianschmutzler replied:
That's what I thought, but is it really worth it?

neheb replied:
I'd say so.

bobafetthotmail replied:
For a script of this size, I doubt it. But that's my habit.

Executing it all once (removing the while loop at the end) aka it's calculating all values on the Helios4 device (in Armbian, the current OS it is running) I get 
time bash -c ./test-script 
./test-script: 90: ./test-script: cannot create /sys/class/hwmon/hwmon1/pwm1: Permission denied
./test-script: 91: ./test-script: cannot create /sys/class/hwmon/hwmon2/pwm1: Permission denied

real	0m0.012s
user	0m0.001s
sys	0m0.012s


Still, what does "easier" mean for your proposed change? 

Should be something like this in shell script (not tested)
```
	for i in {1..10} ; do
	if [ $cpu_temp -le $(( cpu_temp_min + ( ( (cpu_temp_max - cpu_temp_min) / 10 ) * i ) )) ]; then
	    continue
	else
	    pwm_value=$(( pwm_min + ( ( 255 - pwm_min ) / 10 * i ) ))
	    break
	fi
```

Imho it's harder to understand at a glance what is going on, and these devices have ample flash space or a SD card, so even if it is longer it does not matter much.

neheb replied:
I run this thing on Armbian. It's quite slow (I probably have no business running docker on this). Any speed improvement is welcome.

bobafetthotmail replied:
> I run this thing on Armbian. It's quite slow (I probably have no business running docker on this). Any speed improvement is welcome.

You mean Armbian on another device, or you have a Helios 4 too?

Because my benchmark above with Helios 4 running Armbian shows running a script with all calculations only needs 0.012 seconds. The device has a pretty decent CPU.

Imho if you have a Armbian device and it runs bad you need to check RAM consumption and maybe disable zram. I have seen enough times that when I use zram on anything (not just on Armbian) and the device does not have enough free memory the devices get stuck in a loop of "compressing and moving to zram" and "decompressing and moving to ram" and this kills any performance.

neheb replied:
I have a Helios 4. I also have a Turris zombis. Both are the same generation CPU at the same clockspeed. Nextcloud on TurrisOS performs faster than on Armbian.

bobafetthotmail commented:
updated with feedback I agree with (i.e. the fan control script is not using a for loop).

neheb commented:
Your commit title is Helios 3. Description mentions 4.

bobafetthotmail commented:
rebased to resolve conflicts

bobafetthotmail commented:
removed other kernel config options that got in the commit by mistake during the rebase

ynezz commented:
> target/linux/mvebu/patches-5.4/562-gpio-mvebu-multiple-pwm-per-gpio-bank.patch

I'm willing to merge this PR without this patch, which needs to be upstreamed. In other words, it means additional maintenance burden during kernel version bumps.

bobafetthotmail commented:
> > target/linux/mvebu/patches-5.4/562-gpio-mvebu-multiple-pwm-per-gpio-bank.patch
> 
> I'm willing to merge this PR without this patch, which needs to be upstreamed. In other words, it means additional maintenance burden during kernel version bumps.

Ok, I'll change the PR to remove that patch so it can be merged. If I mange to upstream it I'll open another PR with that patch as a backport or something.

bobafetthotmail committed:
mvebu: add Kobol Helios 4 device

The Helios 4 is a NAS from Kobol
that is powered by an Armada 38x
MicroSOM from Solidrun, similarly
to Clearfog.

This device has:
-Armada 38x CPU
 (dual core ARMv7 1.6 Ghz)
-2 GB of ECC RAM
-Gigabit ethernet (Marvell)
-2x USB 3.0 ports
-4x Sata 3.0 ports
-i2c header (J9 |>GND|SDA|SCL|VCC)
-2x 3-pin fan headers with PWM
-micro-usb port is a TTL/UART to
 USB converter connected to TTL
-MicroSD card slot
-System, 4xSata and 1xUSB LEDs

NOT WORKING: fan control

Fan Control requires a kernel patch
that is available in the Armbian
project (the "default firmware"
of this device) and named
mvebu-gpio-remove-hardcoded
-timer-assignment
This patch isn't acceptable
by OpenWrt, it should be upstreamed.
I also have that patch in my own
local OpenWrt builds,
in case you want a more
clean and less confusing patch
for upstreaming.

To install, write the disk image
on a micro SD card with dd or
win32 disk imager, insert the
card in the slot.
Check that the dip switch battery
for boot selection is as follows
Switch 1 and 2 down/off, switches
3, 4, 5 up/on.

Signed-off-by: Alberto Bursi <bobafetthotmail@gmail.com>

bobafetthotmail commented:
@ynezz updated, now the patch (and the fan control script) is removed.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

neheb commented:
@bobafetthotmail can you comment on https://github.com/armbian/build/blob/master/patch/kernel/mvebu-dev/95-helios4-dts-assign-pinctrl-to-fan-and-led.patch ?

neheb commented:
@bobafetthotmail know if this supports > 25MB/s over SD?
